python/bind_data_access.h

```diff
@@ -266,32 +271,71 @@ template <class... Ts> class as_ElementArrayViewImpl {
       return remove_variances(view);
     if (!view.hasVariances())
       init_variances(view);
-    set(view.dims(), get<get_variances>(view), obj);
+    set(view.dims(), view.unit(), get<get_variances>(view), obj);
   }
 
+private:
+  // Helper function object to get a scalar value or variance.
+  template <class View> struct GetScalarVisitor {
+    py::object &self; // The object we're getting the value / variance from.
+    std::remove_reference_t<View> &view; // self as a view.
+
+    template <class Data> auto operator()(const Data &&data) const {
+      if constexpr (std::is_same_v<std::decay_t<decltype(data[0])>,
+                                   scipp::python::PyObject>) {
+        return data[0].to_pybind();
+      } else if constexpr (is_view_v<std::decay_t<decltype(data[0])>>) {
+        auto ret = py::cast(data[0], py::return_value_policy::move);
+        pybind11::detail::keep_alive_impl(ret, self);
+        return ret;
+      } else if constexpr (std::is_same_v<std::decay_t<decltype(data[0])>,
+                                          core::time_point>) {
+        static const auto np_datetime64 =
+            py::module::import("numpy").attr("datetime64");
+        // TODO remove if / when to_string(Unit) supports ASCII only output.
+        const auto unit_str =
+            view.unit() != units::us ? to_string(view.unit()) : "us";
```

Comment:
Since you are using this twice, instead of writing `TODO` comments, just add a helper on top of this file
```cpp
auto ascii_unit(const units::Unit &u) {
  return u == units::us ? "us" : to_string(u);
}
```